lib/src/driver/driver.dart

```diff
@@ -0,0 +1,151 @@
+  /// The maximum number of idle workers at any given time.
+  final int _maxIdleWorkers;
+
+  /// The maximum number of concurrent workers to run at any given time.
+  final int _maxWorkers;
+
+  /// The number of currently active workers.
+  int get _numWorkers => _allWorkers.length;
+
+  /// Idle worker processes.
+  final _idleWorkers = <Process>[];
+
+  /// All workers, even the ones that are in the process of being spawned.
+  final _allWorkers = <FutureOr<Process>>[];
+
+  /// Work requests that haven't been started yet.
+  final _workQueue = new Queue<WorkRequest>();
+
+  /// Factory method that spawns a worker process.
+  final SpawnWorker _spawnWorker;
+
+  BazelWorkerDriver(this._spawnWorker, {int maxIdleWorkers, int maxWorkers})
+      : this._maxIdleWorkers = maxIdleWorkers ?? 4,
+        this._maxWorkers = maxWorkers ?? 4;
+
+  Future<WorkResponse> doWork(WorkRequest request) {
+    var responseCompleter = new Completer<WorkResponse>();
+    _responseCompleters[request] = responseCompleter;
+    _workQueue.add(request);
+    _runWorkQueue();
+    return responseCompleter.future;
+  }
+
+  /// Calls `kill` on all worker processes.
+  Future terminateWorkers() async {
+    var existing = new List.from(_allWorkers);
+    _allWorkers.clear();
+    await Future.wait(existing.map((futureWorker) async {
+      (await futureWorker).kill();
+    }));
+  }
+
+  /// Runs as many items in [_workQueue] as possible given the number of
+  /// available workers.
+  ///
+  /// Will spawn additional workers until [_maxWorkers] has been reached.
+  ///
+  /// This method synchronously drains the [_workQueue] and [_idleWorkers], but
+  /// some tasks may not actually start right away if they need to wait for a
+  /// worker to spin up.
+  void _runWorkQueue() {
+    // Bail out conditions, we will continue to call ourselves indefinitely
+    // until one of these is met.
+    if (_workQueue.isEmpty) return;
+    if (_numWorkers == _maxWorkers && _idleWorkers.isEmpty) return;
+    if (_numWorkers > _maxWorkers) {
+      throw new StateError('Internal error, created to many workers. Please '
+          'file a bug at https://github.com/dart-lang/bazel_worker/issues/new');
+    }
+
+    // At this point we definitely want to run a task, we just need to decide
+    // whether or not we need to start up a new worker.
+    var request = _workQueue.removeFirst();
+    if (_idleWorkers.isNotEmpty) {
+      _runWorker(_idleWorkers.removeLast(), request);
+    } else {
+      // No need to block here, we want to continue to synchronously drain the
+      // work queue.
+      var futureWorker = _spawnWorker();
+      _allWorkers.add(futureWorker);
+      futureWorker.then((worker) {
+        // Somewhat ugly cleanup, we want to replace the `Future<Process>` in
+        // `_allWorkers` with the real `Process` once we have it.
+        _allWorkers.remove(futureWorker);
+        _allWorkers.add(worker);
+
+        // Set up the connection and run the worker.
```

Comment:
Some of these comments are "what" rather than "why". We should see if we can refactor to make it more clear without a comment.